### PR TITLE
[oneDPL] one group radix_sort: + fix with usage of std::forward

### PR DESCRIPTION
**Pull request by MikeDvorskiy**
[oneDPL] one group radix_sort: + fix with usage of std::forward; + usage of 1024 wg size

### CODE REVIEW

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h [hunk 1]
```diff
@@ -702,47 +702,43 @@ __parallel_radix_sort(_ExecutionPolicy&& __exec, _Range&& __in_rng)
 #if __SYCL_COMPILER_VERSION >= 20230101 //for Intel(R) oneAPI C++ Compiler Classic 2023 and later
     //TODO: 1.to reduce number of the kernels; 2.to define work group size in runtime, depending on number of elements
     constexpr auto __wg_size = 64;
-    if (__n <= 32768 && __wg_size * 8 <= __max_wg_size)
+    if (__n <= 32768 && __wg_size * 16 <= __max_wg_size)
```
Comment (dmitriy-sobolev) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:705]: What is the reason to change work_group size?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h [hunk 2]
```diff
@@ -702,47 +702,43 @@ __parallel_radix_sort(_ExecutionPolicy&& __exec, _Range&& __in_rng)
...
-                                            __is_ascending>{}(__exec.queue(), __in_rng);
+            __event = __subgroup_radix_sort<__i_kernel_name<_RadixSortKernel, 8>, __wg_size * 8, 32, __radix_bits,
+                                            __is_ascending>{}(__exec.queue(), ::std::forward<_Range>(__in_rng));
         else
         {
             assert(__n <= 32768);
-            __event = __subgroup_radix_sort<__i_kernel_name<_RadixSortKernel, 7>, __wg_size * 8, 64, __radix_bits,
-                                            __is_ascending>{}(__exec.queue(), __in_rng);
+            __event = __subgroup_radix_sort<__i_kernel_name<_RadixSortKernel, 9>, __wg_size * 16, 32, __radix_bits,
+                                            __is_ascending>{}(__exec.queue(), ::std::forward<_Range>(__in_rng), __proj);
```
Comment (dmitriy-sobolev) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:741]: I guess `__proj` is left here by mistake.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h (hunk 1 above)
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:705]: to use one group radix sort for 16K-32K as well.
It gives better result as the general  pattern.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h (hunk 2 above)
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:741]: yes

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h (hunk 1 above)
Comment (dmitriy-sobolev) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:705]: I see that 16K-32K case is was supported as well. The change doubles work-group size and halves block size (aka number of elements per work-item).

Comment (dmitriy-sobolev) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:705]: I have a concern about that condition `if (__n <= 32768 && __wg_size * 16 <= __max_wg_size)` and to be more precise about `__wg_size * 16 <= __max_wg_size` part. You limit usage of single-work-group algorithm to devices with max work-group size 1024 and more. However, it is only needed for 16k-32k case.

**Comment by dmitriy-sobolev** [comment]
Let's fix other problems here, which break compatibility with older DPC++ compilers:

- `sycl::no_init`; please, use a `__dpl_sycl::__no_init`
- `sycl::plus`; please, use a wrapper `__dpl_sycl::__plus`
- warning: unknown attribute 'reqd_sub_group_size' ignored [-Wunknown-attributes]. Use `intel::reqd_sub_group_size` with the older compiler versions (2021.3 and older).

**Comment by dmitriy-sobolev** [comment]
I also have the following comment for the code. These are optional for this PR and are rather about maintainability.

- Could add comment which tell the reason for having this condition: `#if __SYCL_COMPILER_VERSION >= 20230101 //for Intel(R) oneAPI C++ Compiler Classic 2023 and later`
- Inclusion of `parallel_backend_sycl_radix_sort_one_wg.h` header and its dependency on `parallel_backend_sycl_radix_sort.h` is very weird and convoluted. It can be fixed by several ways, e.g. removing `parallel_backend_sycl_radix_sort_one_wg.h` and injecting its code inside `parallel_backend_sycl_radix_sort.h` or resolving the dependencies by moving then into a new/existing utilities header.

**Comment by MikeDvorskiy** [comment]
> I also have the following comments for the code. These are optional for this PR and are rather about maintainability.
> 
> * Could add comment which tells the reason for having such condition: `#if __SYCL_COMPILER_VERSION >= 20230101 //for Intel(R) oneAPI C++ Compiler Classic 2023 and later`
> * Inclusion of `parallel_backend_sycl_radix_sort_one_wg.h` header and its dependency on `parallel_backend_sycl_radix_sort.h` is very weird and convoluted. It can be fixed by several ways, e.g. removing `parallel_backend_sycl_radix_sort_one_wg.h` and injecting its code inside `parallel_backend_sycl_radix_sort.h` or resolving the dependencies by moving then into a new/existing utilities header.

https://github.com/oneapi-src/oneDPL/issues/804

File: include/oneapi/dpl/pstl/hetero/dpcpp/sycl_defs.h [hunk 3]
```diff
@@ -53,6 +53,12 @@
 #define _ONEDPL_SYCL2023_ATOMIC_REF_PRESENT (_ONEDPL_LIBSYCL_VERSION >= 50500)
 #define _ONEDPL_SYCL_SUB_GROUP_MASK_PRESENT (SYCL_EXT_ONEAPI_SUB_GROUP_MASK == 1) && (_ONEDPL_LIBSYCL_VERSION >= 50700)
 
+#if _ONEDPL_LIBSYCL_VERSION >= 50300
+#    define _ONEDPL_SYCL_REQD_SUB_GROUP_SIZE(SIZE) sycl::reqd_sub_group_size(SIZE)
+#else
+#    define _ONEDPL_SYCL_REQD_SUB_GROUP_SIZE(SIZE)
```
Comment (dmitriy-sobolev) [inline include/oneapi/dpl/pstl/hetero/dpcpp/sycl_defs.h:59]: Should it by rather:
```suggestion
#    define _ONEDPL_SYCL_REQD_SUB_GROUP_SIZE(SIZE) intel::reqd_sub_group_size(SIZE)
```

**Review by dmitriy-sobolev — DISMISSED** [review]
Looks good. However, it may cause conflicts with #783.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h [hunk 4]
```diff
@@ -702,49 +702,40 @@ __parallel_radix_sort(_ExecutionPolicy&& __exec, _Range&& __in_rng)
...
-            __event = __subgroup_radix_sort<__i_kernel_name<_RadixSortKernel, 6>, __wg_size * 8, 32, __radix_bits,
-                                            __is_ascending>{}(__exec.queue(), __in_rng);
-        else
-        {
-            assert(__n <= 32768);
-            __event = __subgroup_radix_sort<__i_kernel_name<_RadixSortKernel, 7>, __wg_size * 8, 64, __radix_bits,
-                                            __is_ascending>{}(__exec.queue(), __in_rng);
-        }
-    }
+    using _RadixSortKernel = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_generator<
```
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:705]: It looks like some lines of code were lost here `_CustomName`, `_RadixBitsType`, `_AscendingType` are all undefined as far as I can tell.

**Review by danhoeflinger — CHANGES_REQUESTED** [review]

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h (hunk 4 above)
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:710]: oops, it seems is "resolving conflict artefact"... Thank you, Dan.

**Review by danhoeflinger — APPROVED** [review]
LGTM

**Pull request merged**